### ISSUE TITLE
Painfully slow analyzing deep array structure with loops

### ISSUE THREAD

**Issue by chrisgraham** [vimeo/psalm#2422]
Analyzing this is super-slow...

https://gitlab.com/composr-foundation/composr/blob/master/sources/webstandards_js_lint.php

I couldn't cut it down to a simple test case as the complexity seems to be apart of it. Looping over the big array structure here, and possibly how it is modified on the fly.

**Comment by muglug** [comment]
Psalm doesn't do well with switch/if statements with broad ranges of possible values (i.e. linters like the one above).

But more broadly, why are you performing linting in PHP when there are excellent (and highly configurable) native JS linters available now?

**Comment by chrisgraham** [comment]
Ok. You can think of this from a user's point of view though - Psalm hangs with no error message or even indicating what file it hangs on, unless you run it in debug mode. I'd suggest at least documenting that if it hangs, run debug mode to see where it is hanging and put in an `ignoreFiles` entry for the file in question.
However better would be if you implemented the ability to stop deep analysis (or whatever you'd call it) on particular files. That way at least it can do some more basic checks and get the API structure of the file. I assume the current `ignoreFiles` implementation just completely skips files.

**Comment by muglug** [comment]
> Ok. You can think of this from a user's point of view though - Psalm hangs with no error message or even indicating what file it hangs on

Yes, that's non-optimal, but only happens on a tiny percentage of (often-badly-written) code. There’s a known OOM, ticketed here (the original linked code has since been improved, but I reproduced it): https://github.com/vimeo/psalm/issues/1943#issuecomment-511382630

> That way at least it can do some more basic checks and get the API structure of the file. I assume the current ignoreFiles implementation just completely skips files.

It doesn’t completely ignore them - it still gathers any docblock information.

**Comment by chrisgraham** [comment]
> > That way at least it can do some more basic checks and get the API structure of the file. I assume the current ignoreFiles implementation just completely skips files.
> 
> It doesn’t completely ignore them - it still gathers any docblock information.

That's good :). Worth documenting perhaps on https://psalm.dev/docs/running_psalm/configuration/#ltprojectfilesgt

**Comment by muglug** [comment]
I've just pushed a fix for the original ticket - mind testing with master to see if it's solved on that file, too?

**Comment by chrisgraham** [comment]
> But more broadly, why are you performing linting in PHP when there are excellent (and highly configurable) native JS linters available now?

I forgot to answer this instance of you going off-topic on your own issue manager :P. I'll answer because you seem to be curious, and because you're a fellow brit emigrant to the US ;-).

The use of our JavaScript linter now is minimal. I don't believe we run it with type checking enabled, meaning the library specification in this file is not actually in active use. Maybe we should get rid of it as the standard JS library has obviously spiraled since we wrote it.

More broadly though:
1) It is very useful to have one streamlined low-dependency system that can do basic checks on everything in one go (our system can check HTML, WCAG, JS, CSS, PHP).
2) A single integrated system is better able to check embedded contexts like <script> tags.
3) Imagine a novice full-stack dev who hasn't gone through your extensive Vimeo training knocking out some quick changes and wants to lint all the files they have worked on before trying to run them without needing to install and know a huge toolchain to do it.

A couple of days ago I also integrated ESLint into our toolchain as an option, but obviously that needs Node installed, not exactly lightweight, so not necessarily going to be there on every system a dev is using.

If starting from scratch I might propose doing it differently, as the duplicated effort would not be worth it. But it's something we have had for a long time and is still useful - so it will be used in conjunction with newer more detailed systems. And that's one reason I'm looking at your PHP linter and other tools at the moment.
